### PR TITLE
feat: add `iter/cusome-by`

### LINKED ISSUE #2338: [RFC]: add `@stdlib/iter/cusome-by`
### Description

This RFC proposes adding the package `@stdlib/iter/cusome-by`, which cumulatively tests whether at least `n` iterated values pass a test implemented by a predicate function. The returned iterator should be a transform iterator, continuing to iterate while source iterator values are available.

```javascript
var array2iterator = require( '@stdlib/array/to-iterator' );

function isPositive( value ) {
	return ( value > 0 );
}

var arr = array2iterator( [ 0, 0, 0, 1, 1 ] );

var it = iterCuSomeBy( arr, 2, isPositive );

var v = it.next().value;
// returns false

v = it.next().value;
// returns false

v = it.next().value;
// returns false

v = it.next().value;
// returns false

v = it.next().value;
// returns true

var bool = it.next().done;
// returns true
```

The `predicate` function should be provided two arguments:

- **value**: the iterated value.
- **index**: iteration index (zero-based).

### Related Issues

No.

### Questions

No.

### Other

- See also `@stdlib/iter/some-by`

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by vivek-anand-singh**
Resolves #2338 

## Description

> What is the purpose of this pull request?
The purpose of this pull request is to introduce the `@stdlib/iter/cusome-by` package, which provides a transform iterator to determine if at least `n` values from a source iterator satisfy a given predicate function. This new feature enhances the library by allowing cumulative testing of iterated values.

This pull request:

-   adds the feature of cusome-by function for `stdib`

## Related Issues

> Does this pull request have any related issues? 
- No

This pull request:

-   resolves #2338
-   fixes #2338

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

### CODE REVIEW

**Comment by vivek-anand-singh** [comment]
@kgryte Please review this PR .

File: lib/node_modules/@stdlib/iter/cusome-by/README.MD [hunk 1]
```diff
@@ -0,0 +1,238 @@
+Here's the adapted README.md for `cusome-by`, based on the `some-by` version:
+
+```markdown
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/README.MD:3]: This is not how to write README files. Furthermore, the README name is incorrect. We don't uppercase `MD`.

File: lib/node_modules/@stdlib/iter/cusome-by/README.MD [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+-   <span class="package-name">[`@stdlib/iter/cuevery-by`][@stdlib/iter/cuevery-by]</span><span class="delimiter">: </span><span class="description">create an iterator which iteratively applies a test function to iterated values, returning `false` if a test fails and `true` otherwise.</span>
+-   <span class="package-name">[`@stdlib/iter/for-each`][@stdlib/iter/for-each]</span><span class="delimiter">: </span><span class="description">create an iterator which invokes a function for each iterated value before returning the iterated value.</span>
+-   <span class="package-name">[`@stdlib/iter/cunone-by`][@stdlib/iter/cunone-by]</span><span class="delimiter">: </span><span class="description">create an iterator which iteratively applies a test function to iterated values, returning `true` if all values fail and `false` otherwise.</span>
+-   <span class="package-name">[`@stdlib/iter/cusome`][@stdlib/iter/cusome]</span><span class="delimiter">: </span><span class="description">create an iterator which iteratively checks whether at least `n` iterated values are truthy.</span>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/README.MD:209]: Do not manually edit this section. See the comment at L197. This section should be empty.

File: lib/node_modules/@stdlib/iter/cusome-by/README.MD [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+[@stdlib/iter/cuevery-by]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/iter/cuevery-by
+
+[@stdlib/iter/for-each]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/iter/for-each
+
+[@stdlib/iter/cunone-by]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/iter/cunone-by
+
+[@stdlib/iter/cusome]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/iter/cusome
+
+<!-- </related-links> -->
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/README.MD:233]: These should be removed.

File: lib/node_modules/@stdlib/iter/cusome-by/benchmark/benchmark.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	if ( typeof count !== 'number' ) {
+		b.fail( 'should be a number' );
+	}
+	b.pass( 'benchmark finished' );
+	b.end();
+
+	function predicate( v ) {
+		return ( v > 0 );
+	}
+});
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/benchmark/benchmark.js:130]: You did not setup EditorConfig. Please read our contributing guidelines.

File: lib/node_modules/@stdlib/iter/cusome-by/docs/repl.txt [hunk 5]
```diff
@@ -0,0 +1,49 @@
+The predicate function is provided two arguments:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/docs/repl.txt:1]: This file does not adhere to project conventions. See the REPL text guide in `/docs`.

File: lib/node_modules/@stdlib/iter/cusome-by/docs/types/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+* @param value - iterated value
+* @param i - iteration index
+* @returns boolean indicating whether an iterated value passes a test
+*/
+type Predicate = Nullary | Unary | Binary;
+
+/**
+* Interface for the returned iterator.
+*/
+interface CuSomeByIterator extends Iterator {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/docs/types/index.d.ts:61]: Why are you doing this? Use `TypedIterator` from `@stdlib/types/iter`. Furthermore, if the source iterator is iterable, the returned iterable should also be iterable. See other `@stdlib/iter/*` packages.

File: lib/node_modules/@stdlib/iter/cusome-by/docs/types/test.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
+import iterCuSomeBy = require( './index' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/docs/types/test.ts:1]: Missing license header. Incorrect indentation throughout the file.

File: lib/node_modules/@stdlib/iter/cusome-by/docs/types/test.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+// The compiler throws an error if the function is provided insufficient arguments...
+{
+    iterCuSomeBy(); // $ExpectError
+    iterCuSomeBy( iterator() ); // $ExpectError
+    iterCuSomeBy( iterator(), 3 ); // $ExpectError
+}
+
+
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/docs/types/test.ts:106]: Why do you have multiple trailing lines?

File: lib/node_modules/@stdlib/iter/cusome-by/examples/index.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+// Perform manual iteration...
+var v;
+while ( true ) {
+	v = it.next();
+	if ( v.done ) {
+		break;
+	}
+	console.log( v.value );
+}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/examples/index.js:45]: EditorConfig.

File: lib/node_modules/@stdlib/iter/cusome-by/lib/index.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+*/
+
+// MODULES //
+
+var main = require( './main.js' );
+
+
+// EXPORTS //
+
+module.exports = main;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/lib/index.js:64]: EditorConfig.

File: lib/node_modules/@stdlib/iter/cusome-by/lib/main.js [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+		if ( FLG ) {
+			return {
+				'done': true
+			};
+		}
+		v = iterator.next();
+		if ( v.done ) {
+			FLG = true;
+			return {
+				'value': false,
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/lib/main.js:118]: You should be assigning `v.value` here.

File: lib/node_modules/@stdlib/iter/cusome-by/lib/main.js [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+				'done': false
+			};
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/lib/main.js:120]:
```suggestion
			return v;
```

File: lib/node_modules/@stdlib/iter/cusome-by/lib/main.js [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+	*/
+	function factory() {
+		return iterCuSomeBy( iterator, n, predicate, thisArg );
+	}
+}
+
+
+// EXPORTS //
+
+module.exports = iterCuSomeBy;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/lib/main.js:152]: EditorConfig.

File: lib/node_modules/@stdlib/iter/cusome-by/package.json [hunk 14]
```diff
@@ -0,0 +1,73 @@
+{
+    "name": "@stdlib/iter/cusome-by",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/package.json:2]: Indentation in this file is off.

File: lib/node_modules/@stdlib/iter/cusome-by/package.json [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+      "iterate",
+      "iterator",
+      "iter",
+      "validate",
+      "predicate",
+      "cumulative",
+      "cusome",
+      "transform"
+    ]
+  }
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/package.json:73]: EditorConfig.

File: lib/node_modules/@stdlib/iter/cusome-by/test/test.js [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+		return function badValue() {
+			iterCuSomeBy( array2iterator( [ 1, 2, 3 ] ), 3, value );
+		};
+	}
+});
+
+tape( 'the function returns an iterator', function test( t ) {
+	var it = iterCuSomeBy( array2iterator( [ 1, 1, 0, 0, 1 ] ), 3, predicate );
+	t.equal( typeof it.next, 'function', 'has next method' );
+	// t.equal( typeof it.return, 'function', 'has return method' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/test/test.js:128]: Don't leave commented code.

File: lib/node_modules/@stdlib/iter/cusome-by/test/test.js [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+	var it;
+	var v;
+	var i;
+
+	values = [ 0, 0, 1, 1, 0, 1 ];
+	expected = [ false, false, false, true, true, true ];
+
+	it = iterCuSomeBy( array2iterator( values ), 2, predicate );
+	t.equal( typeof it.next, 'function', 'has next method' );
+	// t.equal( typeof it.return, 'function', 'has return method' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/test/test.js:148]: Same thing.

File: lib/node_modules/@stdlib/iter/cusome-by/test/test.js [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+
+	r = it.next();
+	t.equal( typeof r.value, 'boolean', 'returns a boolean' );
+	t.equal( r.done, false, 'returns expected value' );
+
+	r = it.next();
+	t.equal( typeof r.value, 'boolean', 'returns a boolean' );
+	t.equal( r.done, false, 'returns expected value' );
+
+	// r = it.return();
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/test/test.js:180]: Why did you do this? Now the test does not test what it is supposed to.

File: lib/node_modules/@stdlib/iter/cusome-by/test/test.js [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+
+	r = it.next();
+	t.equal( typeof r.value, 'boolean', 'returns a boolean' );
+	t.equal( r.done, false, 'returns expected value' );
+
+	r = it.next();
+	t.equal( typeof r.value, 'boolean', 'returns a boolean' );
+	t.equal( r.done, false, 'returns expected value' );
+
+	// r = it.return( 'finished' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/test/test.js:209]: Same thing.

File: lib/node_modules/@stdlib/iter/cusome-by/test/test.js [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+	}
+
+	t.strictEqual( ctx.count, 5, 'returns expected value' );
+	t.end();
+
+	function predicate( v ) {
+		this.count += 1; // eslint-disable-line no-invalid-this
+		return ( v > 0 );
+	}
+});
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/cusome-by/test/test.js:289]: EditorConfig.

File: yarn.lock
Comment (kgryte) [inline yarn.lock:1]: Why did you include this file? Remove.

**Review by kgryte — CHANGES_REQUESTED** [review]
This PR needs significant changes before it can be considered for further review. Please study existing packages in `@stdlib/iter/*` and emulate the style and conventions as closely as possible and as appropriate. And further, please be sure to setup your local dev environment for contributing to stdlib as discussed in the contributing guidelines. It is clear that linting was not setup locally.

**Comment by kgryte** [comment]
@vivek-anand-singh Also, in your OP, you removed the checkbox indicating that you have read the contributing guidelines. Without that checkbox, we cannot accept your PR. Please be sure to include that before any further review.

**Pull request closed without merging**